cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
I have a PR up to harden the string parsing here: https://github.com/apache/arrow/pull/708. Would you mind posting your MSVC version? I'd like to check to make sure that my changes address this issue.